cirq-google/cirq_google/devices/grid_device.py

```diff
@@ -400,25 +426,25 @@ def _value_equality_values_(self):
 def _set_gate_in_gate_spec(
     gate_spec: v2.device_pb2.GateSpecification, gate_family: cirq.GateFamily
 ) -> None:
-    if gate_family == SYC_GATE_FAMILY:
+    if gate_family == _SYC_GATE_FAMILY:
```

Comment:
Your `_build_gateset_and_gate_durations` method inserts a `_SYC_FSIM_GATE_FAMILY` into your gateset if the gate specification proto contains an entry for `syc`; but when converting from a gateset to a gate specification proto; you check whether the gate family in your gateset is `_SYC_GATE_FAMILY` (instead of `_SYC_FSIM_GATE_FAMILY`) to set the `syc` field in the proto. 

Isn't this wrong? or am I missing something?

Reply:
Also, do we have proto -> gateset -> proto roundtrip tests which test this behavior?

Reply:
`GridDevice` cannot be serialized back into `DeviceSpecification` currently. I plan to address this after 1.0 by adding a `to_device_specification()` method which either returns a cached `DeviceSpecification` or converts `GridDeviceMetadata`, TBD.

`create_device_specification_proto()` doesn't serialize a `GridDevice`. It accepts just enough device information to populate the `DeviceSpecification` proto, which doesn't contains some of the generated information in GridDevice like target gatesets. The goal was to simplify the population of `DeviceSpecification` by not having to manually construct the full GridDevice on server side. Because this function doesn't serialize `GridDevice`, this doesn't need to be symmetric with logic in `from_proto`, and instead should be simpler. This is why the gate families for syc are different. For `create_device_specification_proto()`, I used the simplest GateFamily (_SYC_GATE_FAMILY) to represent the syc gate field in `DeviceSpecification` proto. For `from_proto()`, `_SYC_FSIM_GATE_FAMILY` has to be used to accept all gates equivalent to syc.

Will add a comment in `create_device_specification_proto()` to clarify that it doesn't serialize GridDevice.

Reply:
Can you given me an example of a user who would call `create_device_specification_proto` and get back the `DeviceSpecification` ? Who is this user? Where will they obtain the `gateset` that this method expects? Can / should they use the `device.metadata.gateset` that's already present on a grid device?

Reply:
One reason you want to generate a `DeviceSpecification` proto from a GridDevice is server-side, when you are using the GridDevice (which was probably generated with server-side configuration) for client-sent circuits, and you need to generate device specification.

That being said, I am not quite sure I understand the different between `create_device_specification_proto()` and `to_device_specification()`.  Is this something we can post-pone discussion until after 1.0?

I propose doing the minimum we need to get to 1.0, so we can get this PR out (even if it means removing methods we will have to add back later).

Reply:
Discussed with @tanujkhattar offline. The current serialization/deserialization API is confusing to enough people that it definitely needs a revisit after 1.0 :), and some things likely need to be deprecated.

Per Tanuj's suggestion offline, I will change `create_device_specification_proto` to accept both `_*_GATE_FAMILY` and `_*_FSIM_GATE_FAMILY`, so that information from a `GridDevice`'s `GridDeviceMetadata` can be passed into it without modification. This should unblock the 1.0 release.

Reply:
cc @maffoo who also had concerns in this area

Reply:
Updated

Reply:
Should we mark it somehow to note that it is an experimental API and may change in the future?  (e.g. can we make it an underscore-prefix semi-private function?)

Reply:
To be clear, my remaining comment is regarding "The current serialization/deserialization API is confusing to enough people that it definitely needs a revisit after 1.0" and I would like to prevent deprecation (or backwards incompatibility) shortly after we declare that we are not going to do that.  I know we left room for cirq_google, but I would like to avoid as much churn as possible.  If the best we can do is add an "EXPERIMENTAL: May change" disclaimer to the docstring, that will do, but I would prefer preemptive action if we can.